pkg/R/sparkR.R

```diff
@@ -121,45 +119,38 @@ sparkR.init <- function(
   if (sparkRExistingPort != "") {
     sparkRBackendPort <- sparkRExistingPort
   } else {
+    # TODO: test the random port and retry, or use httpuv:::startServer
+    callbackPort = sample(40000, 1) + 10000
     if (Sys.getenv("SPARKR_USE_SPARK_SUBMIT", "") == "") {
       launchBackend(classPath = cp,
                     mainClass = "edu.berkeley.cs.amplab.sparkr.SparkRBackend",
-                    args = as.character(sparkRBackendPort),
+                    args = as.character(callbackPort),
                     javaOpts = paste("-Xmx", sparkMem, sep = ""))
     } else {
       # TODO: We should deprecate sparkJars and ask users to add it to the
       # command line (using --jars) which is picked up by SparkSubmit
       launchBackendSparkSubmit(
           mainClass = "edu.berkeley.cs.amplab.sparkr.SparkRBackend",
-          args = as.character(sparkRBackendPort),
+          args = as.character(callbackPort),
           appJar = .sparkREnv$assemblyJarPath,
           sparkHome = sparkHome,
           sparkSubmitOpts = Sys.getenv("SPARKR_SUBMIT_ARGS", ""))
     }
+    sock <- socketConnection(port = callbackPort, server = TRUE, open = 'rb',
```

Comment:
stdout and stderr is not safe to communicate, any logging will introduce troubles.

--  
Davies Liu
Sent with Sparrow (http://www.sparrowmailapp.com/?sig)


On Thursday, February 26, 2015 at 7:08 PM, Shivaram Venkataraman wrote:

> In pkg/R/sparkR.R (https://github.com/amplab-extras/SparkR-pkg/pull/178#discussion_r25485610):
> > appJar = .sparkREnv$assemblyJarPath, > sparkHome = sparkHome, > sparkSubmitOpts = Sys.getenv("SPARKR_SUBMIT_ARGS", "")) > } > + sock <- socketConnection(port = callbackPort, server = TRUE, open = 'rb',  
> Hmm I see -- Yeah, so R is single-threaded completely and thats why I guess it blocks for a connection to come up.
> BTW, can we use the JVM process's stdoutor a known filename instead ? I know that recently in PySpark you went from stdout to a socket based solution -- was there any reasons for that ?  
>  
> —
> Reply to this email directly or view it on GitHub (https://github.com/amplab-extras/SparkR-pkg/pull/178/files#r25485610).
>  
>  
>  
>  
>